cloudinit/config/cc_chef.py

```diff
@@ -6,78 +6,286 @@
 import json
 import os
+from textwrap import dedent
 
+from cloudinit.config.schema import (
+    get_schema_doc, validate_cloudconfig_schema)
 from cloudinit import templater
 from cloudinit import url_helper
 from cloudinit import util
+from cloudinit.settings import PER_ALWAYS
+
+
+frequency = PER_ALWAYS
+distros = ["all"]
+schema = {
+    'id': 'cc_chef',
+    'name': 'Chef',
+    'title': 'module that configures, starts and installs chef',
+    'description': dedent("""\
+        This module enables chef to be installed (from packages or
+        from gems, or from omnibus). Before this occurs chef configurations are
+        written to disk (validation.pem, client.pem, firstboot.json,
+        client.rb), and needed chef folders/directories are created
+        (/etc/chef and /var/log/chef and so-on). Then once installing proceeds
+        correctly if configured chef will be started (in daemon mode or in
+        non-daemon mode) and then once that has finished (if ran in non-daemon
+        mode this will be when chef finishes converging, if ran in daemon mode
+        then no further actions are possible since chef will have forked into
+        its own process) then a post run function can be executed, which
+        perform finishing activities (such as removing the validation pem
+        file)."""),
+    'distros': distros,
+    'examples': [dedent("""
+        chef:
+          directories:
+            - /etc/chef
+            - /var/log/chef
+          validation_cert: system
+          install_type: omnibus
+          initial_attributes:
+            apache:
+              prefork:
+                maxclients: 100
+              keepalive: off
+          run_list:
+            - recipe[apache2]
+            - role[db]
+          encrypted_data_bag_secret: /etc/chef/encrypted_data_bag_secret
+          environment: _default
+          log_level: :auto
+          omnibus_url_retries: 2
+          server_url: https://chef.yourorg.com:4000
+          ssl_verify_mode: :verify_peer
+          validation_name: yourorg-validator""")],
+    'frequency': frequency,
+    'type': 'object',
+    'properties': {
+        'chef': {
+            'type': 'object',
+            'additionalProperties': False,
+            'properties': {
+                'directories': {
+                    'type': 'array',
+                    'items': {
+                        'type': 'string'
+                    },
+                    'uniqueItems': True,
+                    'description': dedent("""\
+                        Create the necessary directories for chef to run. By
+                        default, it creates the following directories:
+
+                            - ``/etc/chef``
+                            - ``/var/log/chef``
+                            - ``/var/lib/chef``
+                            - ``/var/cache/chef``
+                            - ``/var/backups/chef``
+                            - ``/var/run/chef``""")
```

Comment:
Let's source default strings  from the actual global variable defined (in case this changes in the future).
Maybe something like this:
```suggestion
{chef_dirs}""".format(chef_dirs="\n".join(              
                            ["    - ``{}``".format(d) for d in CHEF_DIRS])))
```